source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.cc

```diff
@@ -154,10 +163,24 @@ void GetAddrInfoDnsResolver::resolveThreadRoutine() {
       if (rc.return_value_ == 0) {
         response = processResponse(*next_query, addrinfo_wrapper.get());
       } else if (reresolve && rc.return_value_ == EAI_AGAIN) {
-        ENVOY_LOG(debug, "retrying query [{}]", next_query->dns_name_);
         absl::MutexLock guard(&mutex_);
-        pending_queries_.push_back(next_query);
-        continue;
+        if (num_retries.has_value()) {
+          (*num_retries)--;
+        }
+        if (num_retries.has_value() && *num_retries > 0) {
+          ENVOY_LOG(debug, "retrying query [{}], num_retries: {}", next_query->dns_name_,
+                    *num_retries);
+          pending_queries_.push_back({std::move(next_query), *num_retries});
+          continue;
+        } else if (num_retries.has_value() && *num_retries == 0) {
+          ENVOY_LOG(debug, "not retrying query [{}] because num_retries: {}", next_query->dns_name_,
+                    *num_retries);
+          response = std::make_pair(ResolutionStatus::Failure, std::list<DnsResponse>());
+        } else {
+          ENVOY_LOG(debug, "retrying query [{}]", next_query->dns_name_);
+          pending_queries_.push_back({next_query, absl::nullopt});
+          continue;
+        }
```

Comment:
I think we could possibly simplify this:
```
        if (!num_retries.has_value()) {
          ENVOY_LOG(debug, "retrying query [{}]", next_query->dns_name_);
          pending_queries_.push_back({next_query, absl::nullopt});
          continue;
        }
        if (*num_retries > 0) {
          ENVOY_LOG(debug, "retrying query [{}], num_retries: {}", next_query->dns_name_,
                    *num_retries);
          pending_queries_.push_back({std::move(next_query), *num_retries});
          continue;
        } 
        ENVOY_LOG(debug, "not retrying query [{}] because num_retries: {}", next_query->dns_name_,
                  *num_retries);
        response = std::make_pair(ResolutionStatus::Failure, std::list<DnsResponse>());
```
This reduces a tiny bit of nesting, and eliminates a redundant call to `num_retries.has_value()`

Reply:
Nice. Thanks! Done.